### PR TITLE
Implement chunking for the build event log

### PR DESCRIPTION
**Pull request by tempoz**
This pull request will implement chunking of blobs to support the build event log

---

**Version bump**: Minor

### CODE REVIEW

File: server/backends/chunkstore/chunkstore.go [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+			continue
+		}
+		w.writeResultChannel <- writeResult{size: bytesFlushed, err: writeError}
+		writeError = nil
+		bytesFlushed = 0
+	}
+	close(w.writeResultChannel)
+}
+
+func (c *Chunkstore) Writer(ctx context.Context, blobName string, blockSize int, timeoutDuration time.Duration) *chunkstoreWriter {
```
Comment (tylerwilliams) [inline server/backends/chunkstore/chunkstore.go:400]: Could we move blockSize and timeoutDuration to the chunkstore constructor -- assuming this is just a library on top of blockstore and we can just instantiate a new chunkstore when we need one? Then Writer can just become:

Writer(ctx context.Context, blobName string)

File: server/backends/chunkstore/chunkstore.go [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+
+func (r *chunkstoreReader) nextChunkExists() (bool, error) {
+	return r.chunkstore.chunkExists(r.ctx, r.blobName, r.nextChunkIndex())
+}
+
+func (r *chunkstoreReader) getNextChunk() error {
+	if exists, err := r.nextChunkExists(); err != nil {
+		return err
+	} else if !exists {
+		return fmt.Errorf("Opening %v: %w", chunkName(r.blobName, r.nextChunkIndex()), os.ErrNotExist)
```
Comment (tylerwilliams) [inline server/backends/chunkstore/chunkstore.go:147]: Can you pick an appropriate util/status error and return those any place we use fmt.Errorfs here? (Prolly status.NotFoundErrorf for this one)

They are handy because they play nicely with gRPC so if any of these errors bubble up uncaught, they'll be displayed/handled correctly in client code. We also have utilities for inspecting these errors and switching on their types which is handy.

File: server/backends/chunkstore/chunkstore.go [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+		offset += r.off + int64(len(r.chunk)-r.chunkOff)
+	default:
+		return r.off, syscall.EINVAL
+	}
+
+	if offset < 0 {
+		return r.off, syscall.EINVAL
+	}
+
+	if offset >= r.off-int64(r.chunkOff) && offset < r.off+int64(len(r.chunk)-r.chunkOff) {
```
Comment (tylerwilliams) [inline server/backends/chunkstore/chunkstore.go:227]: A lot of this chunk /offset math is hard to follow -- I don't think it's wrong necessarily but I also don't have the brainpower to really read this code and check the correctness either :P One way to make it easier to follow would be to add comments any time you modify or check offsets like this?

File: server/backends/chunkstore/chunkstore.go [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+		return result.err
+	case <-time.After(w.timeoutDuration):
+		return os.ErrDeadlineExceeded
+	}
+}
+
+func (w *chunkstoreWriter) writeLoop() {
+	chunkIndex := uint16(0)
+	chunk := []byte{}
+	never := time.Unix(1<<63-62135596801, 999999999)
```
Comment (tylerwilliams) [inline server/backends/chunkstore/chunkstore.go:339]: could you make this a const? or maybe we can remove it entirely? Seems ok if a select case fires that checks for a flush but we don't actually flush anything because there's nothing ready.

File: server/backends/chunkstore/chunkstore_test.go [hunk 5]
```diff
@@ -0,0 +1,416 @@
...
+}
+
+func TestWriter(t *testing.T) {
+	m := mockstore.New()
+	c := New(m)
+	mtx := &mockstore.Mocktext{}
+
+	test_map := make(map[string][]byte)
+
+	w := c.Writer(mtx, "foo", 5, 50*time.Millisecond)
```
Comment (tylerwilliams) [inline server/backends/chunkstore/chunkstore_test.go:374]: nit: maybe define flushTime := 50*time.Millisecond, then you can do:

c.Writer(mtx, "foo", 5, flushTime)
time.Sleep(flushTime + time.Millisecond)
cmp.Equal(...)

File: server/backends/chunkstore/chunkstore.go [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+			flushTime = time.Now().Add(w.timeoutDuration)
+		}
+		if timeout {
+			continue
+		}
+		w.writeResultChannel <- writeResult{size: bytesFlushed, err: writeError}
+		writeError = nil
+		bytesFlushed = 0
+	}
+	close(w.writeResultChannel)
```
Comment (tylerwilliams) [inline server/backends/chunkstore/chunkstore.go:397]: There's a lot of logic here -- could see move some of these blocks into functions and then have this writeLoop just be the control flow?

**Review by tylerwilliams — CHANGES_REQUESTED** [review]
Apologies this has taken me a minute to get to! I think the API mostly looks good and I believe this works but the code is also a lot to grok. I have a few ideas about simplifying it and then some other boring comments about go-style etc. LMK if you want to chat about any of my comments.

File: server/testutil/mockstore/mockstore.go [hunk 7]
```diff
@@ -0,0 +1,54 @@
+package mockstore
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+)
+
+type Mocktext struct{}
```
[the remark this answers is not in the document]
Comment (tempoz) [reply server/testutil/mockstore/mockstore.go:10]: lol, sure, no problem.

File: server/backends/chunkstore/chunkstore.go (hunk 1 above)
Comment (tempoz) [reply server/backends/chunkstore/chunkstore.go:400]: Fine with me!

File: server/backends/chunkstore/chunkstore.go (hunk 3 above)
Comment (tempoz) [reply server/backends/chunkstore/chunkstore.go:227]: Good point; it's easy to get used to it and forget that this stuff is super obtuse.

File: server/backends/chunkstore/chunkstore_test.go (hunk 5 above)
Comment (tempoz) [reply server/backends/chunkstore/chunkstore_test.go:374]: I admit to not having a perfect understanding of goroutines, but I'd be concerned about a test with only a millisecond buffer after another thread is supposed to have timed out might flake a lot. I have no problem with encapsulating the flush time in a variable, though.

File: server/backends/chunkstore/chunkstore.go (hunk 4 above)
Comment (tempoz) [reply server/backends/chunkstore/chunkstore.go:339]: I can make the number a const, but not the `time.Time` (due to it not being a "character, string, boolean, or numeric value"), I tried :P Still, probably worth consting the number if nothing else. I was doing the thing you described before, where it would periodically flush nothing every two minutes, but we talked about it and you pointed out that since we never do anything when it wakes up that it shouldn't wake up, hence this solution. If you prefer the old method (maybe with a longer delay than two minutes, but not the maximum possible time representable in go?), I would be happy to do that instead.

File: server/backends/chunkstore/chunkstore.go [hunk 8]
```diff
@@ -0,0 +1,413 @@
...
+}
+
+func (c *Chunkstore) BlobExists(ctx context.Context, blobName string) (bool, error) {
+	return c.chunkExists(ctx, blobName, 0)
+}
+
+func (c *Chunkstore) ReadBlob(ctx context.Context, blobName string) ([]byte, error) {
+	if data, err := io.ReadAll(c.Reader(ctx, blobName)); err != nil {
+		return []byte{}, err
+	} else {
```
[the remark this answers is not in the document]
Comment (tempoz) [reply server/backends/chunkstore/chunkstore.go:35]: got it. fingers crossed that i remember that in the future.

File: server/backends/chunkstore/chunkstore.go [hunk 9]
```diff
@@ -0,0 +1,413 @@
...
+	"time"
+
+	"github.com/buildbuddy-io/buildbuddy/server/interfaces"
+)
+
+const (
+	mb = 1 << 20
+)
+
+type Chunkstore struct {
```
[the remark this answers is not in the document]
Comment (tempoz) [reply server/backends/chunkstore/chunkstore.go:20]: Yep, np

File: server/backends/chunkstore/chunkstore.go (hunk 2 above)
Comment (tempoz) [reply server/backends/chunkstore/chunkstore.go]: I actually didn't realize buildbuddy had its own error system until I read this comment; I much prefer this model to the standard golang model where errors rarely contain any state at all and wrapping is skin-deep.

File: server/backends/chunkstore/chunkstore.go [hunk 10]
```diff
@@ -0,0 +1,454 @@
...
+
+type ChunkstoreOptions struct {
+	WriteBlockSize       int
+	WriteTimeoutDuration time.Duration
+}
+
+func New(blobstore interfaces.Blobstore, co *ChunkstoreOptions) *Chunkstore {
+	writeBlockSize := co.WriteBlockSize
+	if writeBlockSize == 0 {
+		writeBlockSize = 1 * mb
```
Comment (tylerwilliams) [inline server/backends/chunkstore/chunkstore.go:36]: for these defaults, it'd be nice to define const defaults at the top of the file and use those if the value is not specd

File: server/backends/chunkstore/chunkstore.go [hunk 11]
```diff
@@ -0,0 +1,454 @@
...
+	}
+	writeTimeoutDuration := co.WriteTimeoutDuration
+	if writeTimeoutDuration == 0 {
+		writeTimeoutDuration = 5 * time.Second
```
Comment (tylerwilliams) [inline server/backends/chunkstore/chunkstore.go:40]: (make a default for this one too)

**Review by tylerwilliams — APPROVED** [review]
LGTM
